segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/extendeds3/ExtendedS3Storage.java

```diff
@@ -0,0 +1,394 @@
+    }
+
+    private Void syncWrite(SegmentHandle handle, long offset, InputStream data, int length) throws StreamSegmentSealedException, BadOffsetException {
+        long traceId = LoggerHelpers.traceEnter(log, "write", handle.getSegmentName(), offset, length);
+
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
+
+        SegmentProperties si = syncGetStreamSegmentInfo(handle.getSegmentName());
+
+        if (si.isSealed()) {
+            throw new StreamSegmentSealedException(handle.getSegmentName());
+        }
+
+        if (si.getLength() < offset) {
+            throw new BadOffsetException(handle.getSegmentName(), si.getLength(), offset);
+        }
+
+        client.putObject(this.config.getBucket(), this.config.getRoot() + handle.getSegmentName(),
+                Range.fromOffsetLength(offset, length), data);
+        LoggerHelpers.traceLeave(log, "write", traceId);
+        return null;
+    }
+
+    private Void syncSeal(SegmentHandle handle) {
+        long traceId = LoggerHelpers.traceEnter(log, "seal", handle.getSegmentName());
+
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
+
+        AccessControlList acl = client.getObjectAcl(config.getBucket(),
+                config.getRoot() + handle.getSegmentName());
+        acl.getGrants().clear();
+        acl.addGrants(new Grant[]{new Grant(new CanonicalUser(config.getAccessKey(), config.getAccessKey()),
+                Permission.READ)});
+
+        client.setObjectAcl(
+                new SetObjectAclRequest(config.getBucket(), config.getRoot() + handle.getSegmentName()).withAcl(acl));
+        LoggerHelpers.traceLeave(log, "seal", traceId);
+        return null;
+    }
+
+    private Void syncConcat(SegmentHandle targetHandle, long offset, String sourceSegment) throws StreamSegmentNotExistsException {
+        long traceId = LoggerHelpers.traceEnter(log, "concat", targetHandle.getSegmentName(),
+                offset, sourceSegment);
+
+        SortedSet<MultipartPartETag> partEtags = new TreeSet<>();
+        String targetPath = config.getRoot() + targetHandle.getSegmentName();
+        String uploadId = client.initiateMultipartUpload(config.getBucket(), targetPath);
+
+        // check whether the target exists
+        if (!syncExists(targetHandle.getSegmentName())) {
+            throw new StreamSegmentNotExistsException(targetHandle.getSegmentName());
+        }
+        // check whether the source is sealed
+        SegmentProperties si = syncGetStreamSegmentInfo(sourceSegment);
+        Preconditions.checkState(si.isSealed(), "source segment must be sealed.");
+
+        //Upload the first part
+        CopyPartRequest copyRequest = new CopyPartRequest(config.getBucket(),
+                targetPath,
+                config.getBucket(),
+                targetPath,
+                uploadId,
+                1).withSourceRange(Range.fromOffsetLength(0, offset));
```

Comment:
Can you clarify this more? "You risk dropping data out if the offset is bad". I do not think that happens in any scenario.